torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp

```diff
@@ -127,6 +128,60 @@ RegisterOperators reg(
            };
          },
          aliasAnalysisSpecialCase()),
+     Operator(
+         "prim::AllocateSlab() -> Storage",
+         [](const Node* node) -> Operation {
+           int64_t total_size = node->i(attr::total_size);
+           auto device_type =
+               static_cast<DeviceType>(node->i(attr::device_type));
+           return [total_size, device_type](Stack* stack) {
+             auto allocator = GetAllocator(device_type);
+             auto slab = c10::Storage(
+                 c10::Storage::use_byte_size_t(),
+                 total_size,
+                 allocator,
+                 /*resizable=*/false);
+             push(stack, std::move(slab));
+           };
+         },
+         aliasAnalysisSpecialCase()),
+     Operator(
+         prim::AllocateTensor,
+         [](const Node* node) -> Operation {
+           int64_t size = node->i(attr::size);
+           int64_t offset = node->i(attr::offset);
+           auto type = node->ty(attr::profiled_type)->expect<TensorType>();
+
+           return [offset, size, type](Stack* stack) {
+             c10::Storage slab;
+             pop(stack, slab);
+             uint8_t* start = static_cast<uint8_t*>(slab.data());
+             void* src = static_cast<void*>(start + offset);
+             at::Tensor temp_tensor = at::from_blob(
+                 src,
+                 *type->sizes().concrete_sizes(),
+                 *type->strides().concrete_sizes(),
+                 at::TensorOptions(*type->device()).dtype(*type->scalarType()));
+             temp_tensor.storage().set_nbytes(size);
+             push(stack, std::move(temp_tensor));
+           };
+         },
+         aliasAnalysisSpecialCase()),
+     Operator(
+         "prim::ReleaseSlab(Storage slab, ...) -> ()",
+         [](Stack* stack) {
+           auto num_inputs = pop(stack).toInt();
+           std::vector<at::Tensor> inputs(num_inputs - 1);
```

Comment:
What is the purpose of taking in Tensors here ? otherwise, as soon as a Tensor has it's last use its ref-count goes to zero and it gets deallocated. What does having the Tensors as inputs provide ?

Reply:
For one thing there's a pass in `CodeImpl` that'll prune unused tensors 

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/interpreter/preprocess_graph.cpp#L211

For another I thought being perfectly explicit about free/destruction would be a good thing (this pertains to your comments further down about ref count)

Reply:
Why is a Tensor being pruned after all its uses a bad thing ?

Reply:
So despite what I have implemented here currently I think the answer to this question is "because otherwise we would get a double free". My reasoning: if you 

1. don't delete the `deleter` at the `Storage` allocation site (as in how `AllocateTensor` is now) of the temp tensor (note I'm talking about the `Storage` abstraction rather than actual memory) .
2. let tensors go out of scope before the end of the run (therefore let `Storage` perform a free)

then you will get a double free/the slab being overwritten by the system memory manager. I'm not 100% sure of this because I haven't closely studied `StorageImpl` but intuitively this is how it should work.

Reply:
Is the storage not also ref-coutned ? what do we think is going to be double free'd, because I dont think the storage will be. we can sync more on this or clear up in follow up as well